gateway/hydrator.go

```diff
@@ -129,19 +129,19 @@ func (eph *EndpointHydrator) run() {
 	eph.isHealthy = eph.getHealthStatus(&successfulServiceChecks)
 }
 
+// performChecks performs a single iteration of QoS checks for a given service,
+// using a worker pool pattern to concurrently check endpoints.
 func (eph *EndpointHydrator) performChecks(serviceID protocol.ServiceID, serviceQoS QoSService) error {
 	logger := eph.Logger.With("service_id", string(serviceID))
 
 	// TODO_FUTURE(@adshmh): support specifying the app(s) used for sending/signing synthetic relay requests by the hydrator.
 	// Passing a nil as the HTTP request, because we assume the Centralized Operation Mode being used by the hydrator, which means there is
 	// no need for specifying a specific app.
-	protocolRequestCtx, err := eph.Protocol.BuildRequestContext(serviceID, nil)
-	if err != nil {
-		logger.Warn().Err(err).Msg("Failed to build a protocol request context")
-		return err
-	}
 
-	uniqueEndpoints, err := protocolRequestCtx.AvailableEndpoints()
+	// Get the list of available endpoints for the service per run.
+	// This is done here to define the full set of hydrator check jobs
+	// that will be executed concurrently, using a worker pool pattern.
+	uniqueEndpoints, err := eph.Protocol.GetEndpoints(serviceID, nil)
```

Comment:
One way in which we can fix this bug without breaking the clear definitions of existing interfaces:

- Introduce a new method on protocol: `BuildRequestContextForEndpoint(serviceID, endpointAddr)`
- This will also remove the need for `mock` endpoint selectors on QoS side.

```go
uniqueEndpoints, err := protocolRequestCtx.AvailableEndpoints()

...
for _, serviceRequestCtx := range requiredQoSChecks {
    for _, targetEndpoint := range uniqueEndpoints {
           protocolRequestCtx, err := eph.Protocol.BuildRequestContextForEndpoint(serviceID, targetEndpoint)
           ...
           
   }
}

```